bouncer/src/repo/manipulator/modelconvertor/import/odaHelper/data_collector_oda.cpp

```diff
@@ -0,0 +1,129 @@
+#include "data_collector_oda.h"
+
+/**
+*  Copyright (C) 2018 3D Repo Ltd
+*
+*  This program is free software: you can redistribute it and/or modify
+*  it under the terms of the GNU Affero General Public License as
+*  published by the Free Software Foundation, either version 3 of the
+*  License, or (at your option) any later version.
+*
+*  This program is distributed in the hope that it will be useful,
+*  but WITHOUT ANY WARRANTY; without even the implied warranty of
+*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+*  GNU Affero General Public License for more details.
+*
+*  You should have received a copy of the GNU Affero General Public License
+*  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <OdaCommon.h>
+#include <OdString.h>
+
+#include <toString.h>
+#include <DgLevelTableRecord.h>
+
+#include "../../../../core/model/bson/repo_bson_factory.h"
+using namespace repo::manipulator::modelconvertor::odaHelper;
+
+void DataCollectorOda::triangleOut(const OdInt32* p3Vertices, const OdGeVector3d* pNormal)
+{
+	const OdGePoint3d*  pVertexDataList = vertexDataList();
+	const OdGeVector3d* pNormals = NULL;
+
+	if ((pVertexDataList + p3Vertices[0]) != (pVertexDataList + p3Vertices[1]) &&
+		(pVertexDataList + p3Vertices[0]) != (pVertexDataList + p3Vertices[2]) &&
+		(pVertexDataList + p3Vertices[1]) != (pVertexDataList + p3Vertices[2]))
+	{
+		std::vector<repo::lib::RepoVector3D64> vertices;
+		for (int i = 0; i < 3; ++i)
+		{
+			vertices.push_back({ pVertexDataList[p3Vertices[i]].x , pVertexDataList[p3Vertices[i]].y, pVertexDataList[p3Vertices[i]].z });
+		}
+		OnTriangleOut(vertices);
```

Comment:
This is a very weird workflow, for a parent class function to call a self declared abstract function in a non abstract function.

It's better for children to override this trangleOut and have a parent helper function to extract the vertices.

```suggestion
void DataCollectorOda::triangleOut(const OdInt32* p3Vertices, const OdGeVector3d* pNormal) {
    auto vertices = convertTo3DRepoVertices();
    collector->addFace(vertices);
}
```
This way, DGN doesn't need to override any functions, and RVT will override this, call `convertTo3DRepoVertices()` and do its extra bits.